src/ripple/protocol/LedgerFormats.h

```diff
@@ -24,76 +24,177 @@
 
 namespace ripple {
 
-/** Ledger entry types.
+/** Identifiers for on-ledger objects.
 
-    These are stored in serialized data.
+    Each ledger object requires a unique type identifier, which is stored
+    within the object itself; this makes it possible to iterate the entire
+    ledger and determine each object's type and verify that the object you
+    retrieved from a given hash matches the expected type.
 
-    @note Changing these values results in a hard fork.
+    @warning The range of valid on-ledger identifiers is between 1 and 32767
+             inclusive. Since these values are stored inside objects stored
+             on the ledger they are part of the protocol. **Changing them
+             should be avoided because without special handling, this will
+             result in a hard fork.**
+
+    @note Values outside this range may be used internally by the code for
+          various purposes, but attempting to use such values to identify
+          on-ledger objects will results in an invariant failure.
+
+    @note When retiring types, the specific values should not be removed but
+          should be marked as [[deprecated]]. This is to avoid accidental
+          reuse of identifiers.
+
+    @todo The C++ language does not enable checking for duplicate values
+          here. If it becomes possible then we should do this.
 
     @ingroup protocol
 */
-// Used as the type of a transaction or the type of a ledger entry.
-enum LedgerEntryType {
-    /** Special type, anything
-        This is used when the type in the Keylet is unknown,
-        such as when building metadata.
-    */
-    ltANY = -3,
+// clang-format off
+enum LedgerEntryType : std::uint16_t
+{
+    /** A ledger object which describes an account.
 
-    /** Special type, anything not a directory
-        This is used when the type in the Keylet is unknown,
-        such as when iterating
-    */
-    ltCHILD = -2,
+        \sa keylet::account
+     */
+    ltACCOUNT_ROOT = 0x0061,
```

Comment:
How about adding a comment for these converted ASCII values pointing out that they were chosen based on an ASCII character? e.g. a trailing comment `// 'a'`?

Reply:
Hmm... I'll have to think about that because I intentionally wanted to move away from the  "ascii characters" thing. In an ideal world, we wouldn't have to deal with any of these, as these values would have had used normal `enum` semantics, so that we could avoid having to explicitly choose & specify the value of each member, which is error prone and unnecessarily verbose.